adds article for preloading optional fonts (#2374)
GoogleChrome/web.dev: src/site/content/en/blog/preload-optional-fonts/index.md

---
title: 'No layout shifting and FOIT with font-display: optional'
subhead: 'Starting in Chrome 82, link rel="preload" and font-display: optional can be combined to remove layout jank completely'
authors:
  - houssein
date: 2020-03-18
hero: hero.jpg
description: |

>   This post covers new optional font loading behavior with preloading in Chrome 82

Reviewer: Writers frequently do something more like this for the description (taking from your opening paragraph:

By optimizing rendering cycles, Chrome 82 eliminates layout shifting when
preloading optional fonts. Combining `<link rel="preload">` with `font-display: optional` is the
most effective way to guarantee jank-free rendering of custom fonts.

tags:
  - post
  - performance
  - fonts
---

{% Aside %}
  In Chrome 82, new font loading improvements have been made to completely eliminate layout shifting and flash of invisible text (FOIT) when optional fonts are preloaded.
{% endAside %}

> By optimizing rendering cycles, Chrome 82 removes any chance of layout shifting occurring when

Reviewer: ```suggestion
By optimizing rendering cycles, Chrome 82 eliminates layout shifting when
```

preloading optional fonts. Combining `<link rel="preload">` with `font-display: optional` is the
most effective way to guarantee no layout jank when rendering custom fonts.

## Font rendering

> Layout shifting, or re-layout, occurs when a resource on a web page changes dynamically resulting in

Reviewer: ```suggestion
Layout shifting, or re-layout, occurs when a resource on a web page changes dynamically, resulting in
```

period. This means that "invisible" text is displayed very briefly before switching to a fallback
font. If the font is not downloaded within 100ms, then the fallback font is used and no swapping
occurs.

<figure class="w-figure">
  <img src="./previous-behavior-fail.png" alt="Diagram showing previous optional font behavior when font fails to load">
  <figcaption class="w-figcaption">Previous <code>font-display: optional</code> behavior in Chrome when font is downloaded <b>after</b> the 100ms block period</figcaption>
</figure>

However, in the case that the font is downloaded before the 100ms block period completes, the custom
font is rendered and used on the page.

<figure class="w-figure">
  <img src="./previous-behavior-pass.png" alt="Diagram showing previous optional font behavior when font loads in time">
  <figcaption class="w-figcaption">Previous <code>font-display: optional</code> behavior in Chrome when font is downloaded <b>before</b> the 100ms block period</figcaption>
</figure>

Chrome re-renders the page **twice** in both instances, regardless of whether the fallback font
is used or if the custom font finishes loading in time. This causes a slight flicker of invisible

> text and, in cases when a new font is rendered, a layout jank that moves some of the page's content.

Reviewer: ```suggestion
text and, in cases when a new font is rendered, layout jank that moves some of the page's content.
```

This occurs easily if the font is stored in the browser's disk cache and can load well before the

> block period is complete (~10ms).

Reviewer: ```suggestion
block period is complete (approximately 10ms).
```

Optimizations have landed In Chrome 82 to entirely remove the first render cycle for optional fonts
that are preloaded with [`<link rel="preload'>`](https://web.dev/codelab-preload-web-fonts/).
Instead, rendering is blocked until the custom font has finished loading or a certain period of time
has passed. This timeout period is currently set at 100ms, but may possibly change in the near
future to optimize performance.

<figure class="w-figure">
  <img src="./new-behavior-fail.png" alt="Diagram showing new preloaded optional font behavior when font fails to load">

>   <figcaption class="w-figcaption">New <code>font-display: optional</code> behavior in Chrome when preloaded and font is downloaded <b>after</b> the 100ms block period (no flash of invisible text)</figcaption>

Reviewer: 'behavior in Chrome when **fonts are** preloaded'

I suggest this addition to the caption. As worded, it sounds like it's Chrome that's preloading.

I'm not sure my wording is technically accurate. I do know the current wording needs to change.

Author: Definitely technically accurate, and better than the current wording :)

Fixed!

</figure>

<figure class="w-figure">
  <img src="./new-behavior-pass.png" alt="Diagram showing new preloaded optional font behavior when font loads in time">

>   <figcaption class="w-figcaption">New <code>font-display: optional</code> behavior in Chrome when preloaded and font is downloaded <b>before</b> the 100ms block period (no flash of invisible text)</figcaption>

Reviewer: This caption has the same issue as the previous one.

</figure>

Preloading optional fonts in Chrome removes the possibility of layout jank and flash of unstyled
text. This matches the required behavior as specified in the

> [W3C CSS specification](https://drafts.csswg.org/css-fonts-4/#valdef-font-face-font-display-optional)

Reviewer: ```suggestion
[CSS Fonts Module Level 4](https://drafts.csswg.org/css-fonts-4/#valdef-font-face-font-display-optional)
```

where optional fonts should never cause re-layout and user agents can instead delay rendering for a
suitable period of time.

Although it is not necessary to preload an optional font, it greatly improves the chance for it to

> load in time before the first render cycle, especially if it is not yet stored in the browser's

Reviewer: ```suggestion
load before the first render cycle, especially if it is not yet stored in the browser's
```

---
title: 'No layout shifting and FOIT with font-display: optional'
subhead: 'Starting in Chrome 82, link rel="preload" and font-display: optional can be combined to remove layout jank completely'
authors:
  - houssein
date: 2020-03-18
hero: hero.jpg
description: | 
  By optimizing rendering cycles, Chrome 82 eliminates layout shifting when
  preloading optional fonts. Combining <link rel="preload"> with font-display: optional is the
  most effective way to guarantee jank-free rendering of custom fonts.
tags:
  - post
  - performance
  - fonts
---

{% Aside %}
  In Chrome 82, new font loading improvements have been made to completely eliminate layout shifting and flash of invisible text (FOIT) when optional fonts are preloaded.
{% endAside %}

By optimizing rendering cycles, Chrome 82 eliminates layout shifting when preloading optional fonts.
Combining `<link rel="preload">` with `font-display: optional` is the most effective way to
guarantee no layout jank when rendering custom fonts.

> ## Font rendering

Reviewer 2: ```suggestion
guarantee no layout jank when rendering custom fonts.

## Browser compatibility {: #compatibility }

Check out MDN's data for up-to-date cross-browser support information:

* [`<link rel="preload">`](https://developer.mozilla.org/docs/Web/HTML/Preloading_content#Browser_compatibility)
* [`font-display`](https://developer.mozilla.org/docs/Web/CSS/@font-face/font-display#Browser_compatibility)

## Font rendering
```

<figure class="w-figure">
  <img src="./previous-behavior-fail.png" alt="Diagram showing previous optional font behavior when font fails to load">
  <figcaption class="w-figcaption">Previous <code>font-display: optional</code> behavior in Chrome when font is downloaded <b>after</b> the 100ms block period</figcaption>
</figure>

However, in the case that the font is downloaded before the 100ms block period completes, the custom
font is rendered and used on the page.

<figure class="w-figure">
  <img src="./previous-behavior-pass.png" alt="Diagram showing previous optional font behavior when font loads in time">
  <figcaption class="w-figcaption">Previous <code>font-display: optional</code> behavior in Chrome when font is downloaded <b>before</b> the 100ms block period</figcaption>
</figure>

Chrome re-renders the page **twice** in both instances, regardless of whether the fallback font
is used or if the custom font finishes loading in time. This causes a slight flicker of invisible
text and, in cases when a new font is rendered, layout jank that moves some of the page's content.
This occurs easily if the font is stored in the browser's disk cache and can load well before the

> block period is complete (approximately 10ms).

Reviewer 3: ```suggestion
This occurs even if the font is stored in the browser's disk cache and can load well before the
block period is complete (approximately 100ms).
```
I think this should be "occurs even" because it's pointing out a problem in the previous implementation? Also, elsewhere in the article it refers to 100ms as the block period?

Author: Ah good catch. Fixed to "occurs even" 👍 

The 10ms is in reference to how long it can take to load from the cache. But it is kind of confusing to mention it right after I re-iterate the block period, so I just removed it :)

</figure>

<figure class="w-figure">
  <img src="./new-behavior-pass.png" alt="Diagram showing new preloaded optional font behavior when font loads in time">
  <figcaption class="w-figcaption">New <code>font-display: optional</code> behavior in Chrome when fonts are preloaded and font is downloaded <b>before</b> the 100ms block period (no flash of invisible text)</figcaption>
</figure>

Preloading optional fonts in Chrome removes the possibility of layout jank and flash of unstyled
text. This matches the required behavior as specified in [CSS Fonts Module Level
4](https://drafts.csswg.org/css-fonts-4/#valdef-font-face-font-display-optional) where optional
fonts should never cause re-layout and user agents can instead delay rendering for a suitable period
of time.

Although it is not necessary to preload an optional font, it greatly improves the chance for it to
load before the first render cycle, especially if it is not yet stored in the browser's
cache.

## Conclusion

> We're interested to hear your experiences preloading optional fonts with these new optimizations in

Reviewer 2: ```suggestion
The Chrome team is interested to hear your experiences preloading optional fonts with these new optimizations in
```
Avoid we because it implies that web.dev's opinion == Chrome's opinion
